### PR TITLE
Webpack5: Don't import a second file while a first one is in flight

### PR DESCRIPTION
**Pull request by tmeasday**
Issue: Lazy compilation can fail if CSF files are imported very quickly, either by navigating or via preloading (https://github.com/storybookjs/storybook/pull/17964)

https://github.com/webpack/webpack/issues/15541

## What I did

Introduce a "pipeline" into `import()` when Lazy Compilation is active. The pipeline will stop us from running `import()` a second time while a first `import()` is in flight. 

## How to test

Turn on lazy compilation in the `react-ts` example:

```
  core: {
    channelOptions: { allowFunction: false, maxDepth: 10 },
    disableTelemetry: true,
    builder: {
      name: '@storybook/builder-webpack5',
      options: {
        lazyCompilation: true,
      },
    },
  },
```

## NOTE

@yannbf this will *not* help with the test runner bug as it simply "debounces" things on the client side, but in the case of the TR, there are multiple clients, so that cannot happen. A true fix is (maybe?) to do a similar pipelining in on the lazy compilation backend in webpack.

### CODE REVIEW

File: lib/builder-webpack5/src/preview/iframe-webpack.config.ts [hunk 1]
```diff
@@ -91,7 +91,9 @@ export default async (options: Options & Record<string, any>): Promise<Configura
     const storiesFilename = 'storybook-stories.js';
     const storiesPath = path.resolve(path.join(workingDir, storiesFilename));
 
-    virtualModuleMapping[storiesPath] = toImportFn(stories);
+    const needPipelinedImport =
+      !!(coreOptions.builder as Webpack5BuilderConfig).options?.lazyCompilation && !isProd;
```
Comment (yannbf) [inline lib/builder-webpack5/src/preview/iframe-webpack.config.ts:95]: Does the builder need to be safe checked here too?

Comment (shilman) [reply lib/builder-webpack5/src/preview/iframe-webpack.config.ts:95]: If this is a 7.0 change then `builder` will be required (either from the user or provided by `framework`). I believe this change already exists on the `future/base` branch FWIW

Comment (tmeasday) [reply lib/builder-webpack5/src/preview/iframe-webpack.config.ts:95]: I copied the code from 

https://github.com/storybookjs/storybook/blob/89830ad76384faeaeb0c19df3cb44232cdde261b/lib/builder-webpack5/src/preview/base-webpack.config.ts#L48-L53

where it isn't being null checked so I assume it is always defined somehow.

Comment (yannbf) [reply lib/builder-webpack5/src/preview/iframe-webpack.config.ts:95]: Alright!

File: lib/builder-webpack5/src/preview/iframe-webpack.config.ts [hunk 2]
```diff
@@ -9,7 +9,7 @@ import TerserWebpackPlugin from 'terser-webpack-plugin';
 import VirtualModulePlugin from 'webpack-virtual-modules';
 import ForkTsCheckerWebpackPlugin from 'fork-ts-checker-webpack-plugin';
 
-import type { Options, CoreConfig } from '@storybook/core-common';
+import type { Options, CoreConfig, Webpack5BuilderConfig } from '@storybook/core-common';
```
Comment (shilman) [inline lib/builder-webpack5/src/preview/iframe-webpack.config.ts:12]: This should be cleaned up. There shouldn't be anything in `core-common` that refers to webpack. We can do it in a separate PR

Comment (tmeasday) [reply lib/builder-webpack5/src/preview/iframe-webpack.config.ts:12]: @shilman the entry point that this PR touches (`to-importFn` etc) is webpack specific, but common to WP4 + 5. Should we also move that somewhere?

Comment (shilman) [reply lib/builder-webpack5/src/preview/iframe-webpack.config.ts:12]: If this is a 7.0 change it should be in `core-webpack`. If it's a 6.x change, I guess it can stay 🤷

Comment (tmeasday) [reply lib/builder-webpack5/src/preview/iframe-webpack.config.ts:12]: Happy to make that refactor in 7.0 otherwise though.

Comment (shilman) [reply lib/builder-webpack5/src/preview/iframe-webpack.config.ts:12]: Let's make the refactor here, release in 7.0, and then potentially backport after we've tested. WDYT?

Comment (tmeasday) [reply lib/builder-webpack5/src/preview/iframe-webpack.config.ts:12]: @shilman I moved the utilities over. I don't know how to move the `Webpack5BuilderConfig` type while keeping:

```
export type BuilderConfig =
  | BuilderName
  | BuilderConfigObject
  | Webpack4BuilderConfig
  | Webpack5BuilderConfig;
```

Comment (shilman) [reply lib/builder-webpack5/src/preview/iframe-webpack.config.ts:12]: How about something like the following.

In `core-common`:

```
export type BuilderConfig =
  | BuilderName
  | BuilderConfigObject
```

In `core-webpack` (or `builder-webpack5`?):

```
import type { BuilderConfig as BaseBuilderConfig } from '@storybook/core-common';

export type BuilderConfig = BaseBuilderConfig | Webpack5BuilderConfig;
```

Comment (tmeasday) [reply lib/builder-webpack5/src/preview/iframe-webpack.config.ts:12]: Well but the problem is that `BuilderConfig` is part of `CoreConfig` which is part of `StorybookConfig` which the user imports when defining their SB. 

```ts
import type { StorybookConfig } from '@storybook/react-webpack5/types`
```

Probably what we should do is parameterize it?

```ts
// In `core-common`
type StorybookConfig<TBuilderConfig = BuilderConfig>

// In `react-webpack5`
import type { StorybookConfig as BaseStorybookConfig } from '@storybook/core-common';
import type { BuilderConfig } from '@storybook/builder-webpack5';

export type StorybookConfig = BaseStorybookConfig<BuilderConfig>;
```

Does that sounds right? WDYT @ndelangen?

Comment (ndelangen) [reply lib/builder-webpack5/src/preview/iframe-webpack.config.ts:12]: Generally the idea is to keep composing the StorybookConfig up into the framework.

https://github.com/storybookjs/storybook/blob/ea5c7a7616fb7f22b13315a3a16f6a8484864583/frameworks/react-webpack5/src/types.ts#L47-L49

https://github.com/storybookjs/storybook/blob/ea5c7a7616fb7f22b13315a3a16f6a8484864583/presets/react-webpack/src/types.ts#L51-L55

https://github.com/storybookjs/storybook/blob/8e84f36d0f3abc75867b6b69de3e1c12c23b9166/lib/core-webpack/src/types.ts#L41-L42

---

But -correct me if I'm wrong- but we expect users to only have to specify the builder config if they are NOT using a framework, right?
I think using a framework AND setting a builder should not really happen?

Placing the builder config 

---

So if the above is true, the main question is: where does the user import the "StorybookConfig" from if they are not using a framework?

I think the answer is: probably the renderer.

But alternatively we ask the user to do the following:

```ts
// user's main.js
import type { StorybookConfig as StorybookConfigBase } from '@storybook/react`;
import type { StorybookConfig as StorybookConfigBuilder } from '@storybook/builder-webpack5`;

type StorybookConfig = StorybookConfigBase & StorybookConfigBuilder


module.exports: StorybookConfig = {
  // ...
};
```

We could do the above in the framework, move this code:
https://github.com/storybookjs/storybook/blob/414d530df317b3b187d4df76380dba9bb27209c1/lib/core-common/src/types.ts#L18-L24
into `builder-webpack5` package, import it from the framework and compose the types in the framework package. That seems to be the correct thing to do.

1 thing is extremely clear though: we should not get into the position where we have cyclical dependencies in our types.

Comment (tmeasday) [reply lib/builder-webpack5/src/preview/iframe-webpack.config.ts:12]: @ndelangen 

> But -correct me if I'm wrong- but we expect users to only have to specify the builder config if they are NOT using a framework, right?
> I think using a framework AND setting a builder should not really happen?

In this case we are talking about the builder options like `lazyCompilation` or `fsCache`. So I don't think your assumption here is quite right.

Comment (ndelangen) [reply lib/builder-webpack5/src/preview/iframe-webpack.config.ts:12]: I think those should be framework options, flowing into the builderOptions then.

I think we want to move these builderOptions into the builder, and adjust the frameworkOptions to take them, and flow them into the builder.

Comment (tmeasday) [reply lib/builder-webpack5/src/preview/iframe-webpack.config.ts:12]: @shilman are you OK with the current state of this PR assuming the above gets sorted in https://linear.app/chromaui/issue/SB-347/sequence-importing-stories-in-parallel-in-webpack5-builder

**Review by shilman — CHANGES_REQUESTED** [review]
CI build step is failing
